static/app/views/stories/storyTree.tsx

```diff
@@ -1,167 +1,203 @@
-    (a, b) => Number(!!Object.keys(b[1]).length) - Number(!!Object.keys(a[1]).length)
-  );
+  if (props.node.expanded !== expanded) {
+    setExpanded(props.node.expanded);
+  }
+
+  if (!props.node.visible) {
+    return null;
+  }
 
   return (
-    <UnorderedList>
-      {entries.map(([name, children]) => {
-        const childPath = toPath(path, name);
-
-        if (Object.keys(children).length === 0) {
-          const isCurrent = childPath === currentFile ? true : undefined;
-          const to = `/stories/?name=${childPath}`;
-          return (
-            <ListItem key={name} aria-current={isCurrent}>
-              <FolderLink to={to}>
-                <IconFile size="xs" />
-                {name}
-              </FolderLink>
-            </ListItem>
-          );
-        }
-
-        return (
-          <ListItem key={name}>
-            <Folder open>
-              <FolderName>{name}</FolderName>
-              <FolderContent path={childPath} content={children} />
-            </Folder>
-          </ListItem>
-        );
-      })}
-    </UnorderedList>
+    <li>
+      <FolderName
+        onClick={() => {
+          props.node.expanded = !props.node.expanded;
+          setExpanded(props.node.expanded);
+        }}
+      >
+        <IconChevron size="xs" direction={expanded ? 'down' : 'right'} />
+        {normalizeFilename(props.node.name)}
+      </FolderName>
+      {expanded && Object.keys(props.node.children).length > 0 && (
+        <StoryList>
+          {Object.values(props.node.children)
+            .sort(folderOrSearchScoreFirst)
+            .map(child => {
+              if (!child.visible) {
+                return null;
+              }
+              return Object.keys(child.children).length === 0 ? (
+                <File key={child.path} node={child} />
+              ) : (
+                <Folder key={child.path} node={child} />
+              );
+            })}
+        </StoryList>
+      )}
+    </li>
   );
 }
 
-interface TreeMapping extends Record<string, TreeMapping> {}
-
-function toTree(files: string[]): TreeMapping {
-  const root: Record<string, TreeMapping> = {};
-  for (const file of files) {
-    const parts = file.split('/');
-    let tree = root;
-    for (const part of parts) {
-      if (!(part in tree)) {
-        tree[part] = {};
-      }
-      tree = tree[part]!;
-    }
-  }
-  return root;
-}
+function File(props: {node: StoryTreeNode}) {
+  const location = useLocation();
+  const query = qs.stringify({...location.query, name: props.node.path});
 
-function toPath(path: string, name: string) {
-  return [path, name].filter(Boolean).join('/');
+  return (
+    <li>
+      <FolderLink
+        to={`/stories/?${query}`}
+        active={location.query.name === props.node.path}
+      >
+        {/* @TODO (JonasBadalic): Do file type icons make sense here? */}
+        <IconFile size="xs" />
+        {/* @TODO (JonasBadalic): Do we need to show the file extension? */}
```

Comment:
IMO nice-to-have for copy-pasting and making it obvious where the file is coming from, but not necessary